cnudie/retrieve.py

```diff
@@ -634,3 +647,88 @@ def greatest_component_versions(
         versions = versions[:versions.index(greatest_version)+1]
 
     return versions[-max_versions:]
+
+
+def greatest_version_before(
+    component_name: str,
+    component_version: str,
+    ctx_repo: cm.RepositoryContext,
+    oci_client: oc.Client=None,
+):
+    if not isinstance(ctx_repo, cm.OciRepositoryContext):
+        raise NotImplementedError(ctx_repo)
+
+    if not oci_client:
+        oci_client = ccc.oci.oci_client()
+
+    versions = component_versions(
+        component_name=component_name,
+        ctx_repo=ctx_repo,
+        oci_client=oci_client,
+    )
+    versions = sorted(versions, key=version.parse_to_semver)
+    versions = [
+        v for v in versions
+        if version.parse_to_semver(v) < version.parse_to_semver(component_version)
+    ]
+    if len(versions) == 0:
+        return None # no release before current was found
+    return versions[-1]
+
+
+def greatest_component_version_with_matching_minor(
+    component_name: str,
+    ctx_repo: cm.RepositoryContext,
+    reference_version: str,
+    ignore_prerelease_versions: bool=False,
+    oci_client: oc.Client=None,
+) -> str:
+    if not isinstance(ctx_repo, cm.OciRepositoryContext):
+        raise NotImplementedError(ctx_repo)
+
+    if not oci_client:
+        oci_client = ccc.oci.oci_client()
+
+    versions = component_versions(
+        component_name=component_name,
+        ctx_repo=ctx_repo,
+        oci_client=oci_client,
+    )
+
+    return version.find_latest_version_with_matching_minor(
+        reference_version=reference_version,
+        versions=versions,
+        ignore_prerelease_versions=ignore_prerelease_versions,
+    )
+
+
+def greatest_component_version_by_name(
+    component_name: str,
+    ctx_repo: cm.RepositoryContext,
+    ignore_prerelease_versions: bool=False,
+    oci_client: oc.Client=None,
+    component_descriptor_lookup: ComponentDescriptorLookupById=None,
+):
```

Comment:
```suggestion
) -> cm.component:
```
not entirely sure if `cm.component` is the correct alias, please check :). But given the name of the function one might expect a `str`, so I think this hint might be useful